cluster/ubuntu/util.sh

```diff
@@ -346,99 +349,229 @@ function kube-up() {
-  echo "Deploying master on machine ${MASTER_IP}"
-  echo
+  
+  echo -e "\nDeploying master on machine ${MASTER_IP}"
+
   ssh $SSH_OPTS $MASTER "mkdir -p ~/kube/default"
-  scp -r $SSH_OPTS saltbase/salt/generate-cert/make-ca-cert.sh ubuntu/reconfDocker.sh ubuntu/config-default.sh ubuntu/util.sh ubuntu/master/* ubuntu/binaries/master/ "${MASTER}:~/kube"
-
-  # remote login to MASTER and use sudo to configue k8s master
-  ssh $SSH_OPTS -t $MASTER "source ~/kube/util.sh; \
-                            setClusterInfo; \
-                            create-etcd-opts; \
-                            create-kube-apiserver-opts "${SERVICE_CLUSTER_IP_RANGE}" "${ADMISSION_CONTROL}" "${SERVICE_NODE_PORT_RANGE}"; \
-                            create-kube-controller-manager-opts "${NODE_IPS}"; \
-                            create-kube-scheduler-opts; \
-                            create-flanneld-opts "127.0.0.1"; \
-                            sudo -p '[sudo] password to start master: ' cp ~/kube/default/* /etc/default/ && sudo cp ~/kube/init_conf/* /etc/init/ && sudo cp ~/kube/init_scripts/* /etc/init.d/ ;\
-                            sudo groupadd -f -r kube-cert; \
-                            ${PROXY_SETTING} sudo -E ~/kube/make-ca-cert.sh ${MASTER_IP} IP:${MASTER_IP},IP:${SERVICE_CLUSTER_IP_RANGE%.*}.1,DNS:kubernetes,DNS:kubernetes.default,DNS:kubernetes.default.svc,DNS:kubernetes.default.svc.cluster.local; \
-                            sudo mkdir -p /opt/bin/ && sudo cp ~/kube/master/* /opt/bin/; \
-                            sudo service etcd start; \
-                            sudo FLANNEL_NET=${FLANNEL_NET} ~/kube/reconfDocker.sh "a";" 
-}
+
+  # copy the binaries and scripts to the ~/kube directory on the master
+  scp -r $SSH_OPTS \
+    saltbase/salt/generate-cert/make-ca-cert.sh \
+    ubuntu/reconfDocker.sh \
+    ubuntu/config-default.sh \
+    ubuntu/util.sh \
+    ubuntu/master/* \
+    ubuntu/binaries/master/ \
+    "${MASTER}:~/kube"
+
+  EXTRA_SANS=(
+    IP:$MASTER_IP
+    IP:${SERVICE_CLUSTER_IP_RANGE%.*}.1
+    DNS:kubernetes
+    DNS:kubernetes.default
+    DNS:kubernetes.default.svc
+    DNS:kubernetes.default.svc.cluster.local
+  )
+  
+  EXTRA_SANS=$(echo ${EXTRA_SANS[@]} | tr ' ' ',')
+
+  # remote login to MASTER and configue k8s master
+  ssh $SSH_OPTS -t ${MASTER} "
+    source ~/kube/util.sh
+
+    setClusterInfo
+    create-etcd-opts
+    create-kube-apiserver-opts \
+      '${SERVICE_CLUSTER_IP_RANGE}' \
+      '${ADMISSION_CONTROL}' \
+      '${SERVICE_NODE_PORT_RANGE}'
+    create-kube-controller-manager-opts '${NODE_IPS}'
+    create-kube-scheduler-opts
+    create-flanneld-opts '127.0.0.1'
+    sudo -E -p '[sudo] password to start master: ' -- /bin/bash -c '
+      cp ~/kube/default/* /etc/default/ 
+      cp ~/kube/init_conf/* /etc/init/ 
+      cp ~/kube/init_scripts/* /etc/init.d/
+      
+      groupadd -f -r kube-cert
+      "${PROXY_SETTING}" ~/kube/make-ca-cert.sh "${MASTER_IP}" "${EXTRA_SANS}"
```

Comment:
Please note that three levels of quoting are being attempted here, but it is not done correctly.
The intended outermost level of quoting begins on line 379 and ends on line 402; this is constructing a long string argument to and SSH command to ${MASTER}.  The intended middle level of quoting begins on line 391 and ends on line 402; it is constructing a multi-line argument to `sudo`.  The intended innermost level of quoting lies entirely on line 397 and occurs three times --- around ${PROXY_SETTING}, ${MASTER_IP}, and ${EXTRA_SANS}.  But the inner dquotes are not escaped properly.  What actually happens is that the outer level of quoting begins on line 379 and ends early, at the start of line 397.  That quoted string is concatenated with the expansion of ${PROXY_SETTING}, then " ~/kube/make-ca-cert.sh ", then the expansion of ${MASTER_IP}, then a space, then the expansion of ${EXTRA_SANS}, then another long string that begins at the end of line 397 and ends at the beginning of line 402.

Here is a demonstration of how a shorter example goes wrong:
mspreitz@ovespa003:~$ X="arg2; cat"
mspreitz@ovespa003:~$ ssh localhost "sudo /bin/bash -c 'echo arg1 "${X}" arg3' " 
arg1 arg2
cat: arg3: No such file or directory